src/main/java/io/pravega/connectors/flink/EventTimeOrderingFunction.java

```diff
@@ -0,0 +1,125 @@
+ */
+package io.pravega.connectors.flink;
+
+import org.apache.flink.api.common.state.MapState;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.ListTypeInfo;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
+import org.apache.flink.util.Collector;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class EventTimeOrderingFunction<T> extends KeyedProcessFunction<String, T, T> {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final String EVENT_QUEUE_STATE_NAME = "eventQueue";
+
+    private static final String LAST_TRIGGERING_TS_STATE_NAME = "lastTriggeringTsState";
+
+    /**
+     * The input type information for buffering events to managed state.
+     */
+    private final TypeInformation<T> typeInformation;
+
+    /**
+     * The queue of input elements keyed by event timestamp.
+     */
+    private transient MapState<Long, List<T>> elementQueueState;
+
+    /**
+     * State to keep the last triggering timestamp. Used to filter late events.
+     */
+    private transient ValueState<Long> lastTriggeringTsState;
+
+    public EventTimeOrderingFunction(TypeInformation<T> typeInformation) {
+        this.typeInformation = typeInformation;
+    }
+
+    @Override
+    public void open(Configuration config) throws Exception {
+        super.open(config);
+
+        // create a map-based queue to buffer input elements
+        if (elementQueueState == null) {
+            MapStateDescriptor<Long, List<T>> elementQueueStateDescriptor = new MapStateDescriptor<>(
+                    EVENT_QUEUE_STATE_NAME,
+                    BasicTypeInfo.LONG_TYPE_INFO,
+                    new ListTypeInfo<>(this.typeInformation)
+            );
+            elementQueueState = getRuntimeContext().getMapState(elementQueueStateDescriptor);
+        }
+
+        // maintain a timestamp so anything before this time will be ignored
+        if (lastTriggeringTsState == null) {
+            ValueStateDescriptor<Long> lastTriggeringTsDescriptor =
+                    new ValueStateDescriptor<>(LAST_TRIGGERING_TS_STATE_NAME, Long.class);
+            lastTriggeringTsState = getRuntimeContext().getState(lastTriggeringTsDescriptor);
+        }
+    }
+
+    @Override
+    public void processElement(T element, Context ctx, Collector<T> out) throws Exception {
+        // timestamp of the processed element
+        Long timestamp = ctx.timestamp();
+        if (timestamp == null) {
+            // elements with no time component are simply forwarded.
+            // likely cause: the time characteristic of the program is not event-time.
```

Comment:
There are processing time, ingestion time and event time here, and only processing time is making the `timestamp` null here, so you can change the comment as 
```
Simply forward the elements when the time characteristic of the program is set to ProcessingTime
```

Reply:
Changed